google-cloud-pubsub/src/main/java/com/google/cloud/pubsub/v1/StreamingSubscriberConnection.java

```diff
@@ -131,9 +149,18 @@ public void onNext(StreamingPullResponse response) {
           new Runnable() {
             @Override
             public void run() {
-              // Only if not shutdown we will request one more batches of messages to be delivered.
-              if (isAlive()) {
-                requestObserver.request(1);
+              // Only request more if we're not shutdown.
+              // If errorFuture is done, the stream has either failed or hung up,
+              // and we don't need to request.
+              if (isAlive() && !errorFuture.isDone()) {
+                lock.lock();
```

Comment:
This should only be a reader lock, as described earlier

Reply:
I think a normal lock here is OK. There's only one thread that can call `request` at a given time, so making it a read lock shouldn't make anything go faster.

Reply:
Why can only one thread call request at a time?

Reply:
The callback is called every message *batch* processed. Eg, if the `StreamingPullResponse` we receive has 20 messages. We'll call the callback once after all 20 has been processed. The `request(1)` here means one response, not one message.

For each connection, we request one response, process all messages, then execute the callback which requests another. So there's no way this method can be executed in parallel.

I ran the loadtest, apart from https://github.com/GoogleCloudPlatform/pubsub/pull/124, we observed no regression.